// @testcontainers/oraclefree/build/oraclefree-container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StartedOracleDbContainer = exports.OracleDbContainer = void 0;
const testcontainers_1 = require("testcontainers");
/** Default Oracle listener port. */
const ORACLEDB_PORT = 1521;
/** Default pluggable database name provided by the Oracle Free image. */
const DEFAULT_DATABASE = "FREEPDB1";
/**
 * Testcontainers wrapper for Oracle Free.
 *
 * Supports configuring application user credentials and optional custom database creation.
 */
class OracleDbContainer extends testcontainers_1.GenericContainer {
    username = "test";
    password = "test";
    database = undefined;
    constructor(image) {
        super(image);
        this.withExposedPorts(ORACLEDB_PORT);
        this.withWaitStrategy(testcontainers_1.Wait.forLogMessage("DATABASE IS READY TO USE!"));
        this.withStartupTimeout(120_000);
    }
    /** Sets the application username created at container startup. */
    withUsername(username) {
        this.username = username;
        return this;
    }
    /** Sets the password for both SYS and application user startup configuration. */
    withPassword(password) {
        this.password = password;
        return this;
    }
    /** Sets a custom application database/service name. */
    withDatabase(database) {
        if (database.trim() === "") {
            throw new Error("Database name cannot be empty.");
        }
        if (database.toUpperCase() === DEFAULT_DATABASE) {
            this.database = undefined;
            return this;
        }
        this.database = database;
        return this;
    }
    /** Starts the container and returns a typed started container instance. */
    async start() {
        this.withEnvironment({
            ORACLE_PASSWORD: this.password,
            APP_USER: this.username,
            APP_USER_PASSWORD: this.password,
        });
        if (this.database) {
            this.withEnvironment({
                ORACLE_DATABASE: this.database,
            });
        }
        return new StartedOracleDbContainer(await super.start(), this.username, this.password, this.database ?? DEFAULT_DATABASE);
    }
}
exports.OracleDbContainer = OracleDbContainer;
/** Represents a running Oracle Free test container with Oracle-specific accessors. */
class StartedOracleDbContainer extends testcontainers_1.AbstractStartedContainer {
    username;
    password;
    database;
    constructor(startedTestContainer, username, password, database) {
        super(startedTestContainer);
        this.username = username;
        this.password = password;
        this.database = database;
    }
    /** Returns the mapped Oracle listener port. */
    getPort() {
        return this.getMappedPort(ORACLEDB_PORT);
    }
    /** Returns the configured application username. */
    getUsername() {
        return this.username;
    }
    /** Returns the configured password. */
    getPassword() {
        return this.password;
    }
    /** Returns the configured service/database name. */
    getDatabase() {
        return this.database;
    }
    /** Returns a host:port/database URL fragment. */
    getUrl() {
        return `${this.getHost()}:${this.getPort()}/${this.database}`;
    }
    /** Returns an Oracle connection descriptor string (TNS format). */
    getConnectionDescriptor() {
        return `(DESCRIPTION=(ADDRESS=(PROTOCOL=TCP)(HOST=${this.getHost()})(PORT=${this.getPort()}))(CONNECT_DATA=(SERVICE_NAME=${this.database})))`;
    }
}
exports.StartedOracleDbContainer = StartedOracleDbContainer;
